### ISSUE TITLE
Lazily fetch the list of linkshere

### ISSUE THREAD

**Issue by abinashmeher999** [abinashmeher999/pymediawiki#29]
The response of `linkshere` is huge and it blocks the execution till it returns. So instead of setting the limit as `max` it would be good to set it to something reasonable and wrap it in a generator.

**Comment by ayuhsya** [comment]
This will be fixed if we resolve #25 :confused:

**Comment by abinashmeher999** [comment]
Maybe I should have worded it differently. Sometimes the user doesn't need the whole bulk of it and should be able to handle whatever has been received so far. Asynchronous GET would still get the whole bulk of the response. IMO this is a little different approach to make things faster.

**Comment by abinashmeher999** [comment]
Once this is done it can be easily combined with asynchronous GET for even better performance.

**Comment by ayuhsya** [comment]
Well, the `get_linkshere` function already has a parameter for` lhlimit`, so the user can limit it, if and when he requires. I don't see how further can we approach this issue.

**Comment by abinashmeher999** [comment]
Say `lhlimit` is set. That only limits the number of entries in the single response and will mark the response with `lhcontinue`. But the current code `get_linkshere` won't stop till it keeps on encountering `lhcontinue`. It is still going to wait till all responses are received and only then it will return the whole array of titles. 

Fixing this won't require much effort only that once it gets the response of one request it `yield`s to the caller of the generator.

**Comment by ayuhsya** [comment]
@abinashmeher999 Bro your use of the word **generator** confuses me.

**Comment by ghostwriternr** [comment]
Hello @ayuhsya, can you tell me which part of his statements you find confusing? I can try going over your doubts, and see if I can help.

In case you haven't worked with Python's `yield` and generators before, you could go over this article here: https://jeffknupp.com/blog/2013/04/07/improve-your-python-yield-and-generators-explained/

It explains the concept in good detail, with some examples as well. Try to skim through it, and let me know in case you still have doubts 😄

**Comment by ayuhsya** [comment]
@ghostwriternr Yes, that's what was confusing me. There are generators ( a way to specify pages in the GET request ) in the MediaWiki too.
Now I get it that he was talking about python generators. :neutral_face:

**Comment by ghostwriternr** [comment]
That's alright. He was indeed talking about Python's 'generator'. Do you now understand how they could be helpful with accomplishing our task here?

Basically, [our function (get_linkshere)](https://github.com/abinashmeher999/pymediawiki/blob/master/sample/core.py#L63) can be used as a generator function, by using `yield` instead of `return` (and some other minor modifications) to give control back to our calling function. The advantage here of course being, after we `yield` a chunk, we only go back and get the next part (through 'continue') if needed. And generator helps with this by saving the state when returning control, so it's easy to simply resume and get the next chunk. 

Does that make sense? If it doesn't try going through the examples in the article I had linked. If it still confuses you, let me know and I can explain it in detail.

**Comment by ghostwriternr** [comment]
You could also think of it as one function being able to `return` many times, and not having to always start over. So we are essentially `return`ing the different chunks one by one using the same function.

**Comment by ayuhsya** [comment]
@ghostwriternr Let's say user asks for 300 links. So in the first call we fetch 300 links and yield it. Now user later asks for 100 links. How do we proceed then? Because the maximum that can be fetched in one call is 500 and if we want 100 new records, we again need to call the API and get those first 500 links and manually extract the records from 301 to 400. And this is before going into 'continue'.

**Comment by ghostwriternr** [comment]
**EDIT: Misinterpreted a few things. Ignore this comment.**

Your query is a very valid use case, @ayuhsya. As @abinashmeher999 has mentioned [here](https://github.com/abinashmeher999/pymediawiki/issues/29#issuecomment-266267371), we would like to break down even this first 'page' of results, and be able to return it in parts. We go back and resume when we need more.

So taking your example, say we set a smaller limit (say `smaller_limit = 100` or `smaller_limit = 250`), and `yield` those many results first. And if the user needs more, our generator function will 'resume' and `yield` the next `smaller_limit` results. So we're never really fetching the same set of results more than once.

I believe we don't (yet) want to give the user the option to choose if he / she wants to fetch the first 300 results, and then the next 100, like the example in your query. That'd deviate a bit from the actual problem we're trying to solve here, though it can certainly be done. Our primary concern now is to be able to return sooner, and not wait for the entire chunk of the result.

**Comment by ghostwriternr** [comment]
Although @abinashmeher999, can you please clarify which of these paths we want to go:
- Return in chunks, but let the user ask for 'more' links
- Return the entire result set (i.e. all the links) in chunks

You've mentioned both approaches in [these](https://github.com/abinashmeher999/pymediawiki/issues/29#issue-194816533) [comments](https://github.com/abinashmeher999/pymediawiki/issues/29#issuecomment-266267129).

**Comment by ayuhsya** [comment]
@ghostwriternr But the thing is if let's say we make an API call with 250 entries. Now, there is no other way to get the next 250 entries, but to again make the same request and manually extract the entries 251 to 500. If I am missing something in the API then please point it out.

**Comment by ayuhsya** [comment]
I think you are trying to compare our case with the `get_prime()` example given in the blog. In the `get_prime()` case the output was being produced one by one.

**Comment by ghostwriternr** [comment]
@ayuhsya, you're totally right. That was a total oversight on my part, which lead to me quoting the wrong example. I wrongly assumed from your [previous query](https://github.com/abinashmeher999/pymediawiki/issues/29#issuecomment-268823039) that this was possible. Thanks for pointing it out.

Here's a better (and correct) explanation. Now as you had [previously mentioned](https://github.com/abinashmeher999/pymediawiki/issues/29#issuecomment-266267274), we can of course reduce `lhlimit` from "max" to a smaller value (say, `smaller_value`). So this will mark every `smaller_value` values with 'continue'. Now, our current code goes on to fetch all subsequent pages as well, and `return`s it all in one go. Instead, once we have obtained a particular page, we can `yield` it and then go back to fetch the next page using 'continue' when needed. We achieve our goal this way.

Does this make sense?

**Comment by abinashmeher999** [comment]
@ayuhsya The user only mentions how many entries does (s)he want to fetch per call to the API and the generator should keep on "continuing" till no more entries are available through the API. So if the user wants to use only the first 300 (s)he can do a `for index, entry in zip(range(0, 300), page.categories(num_per_fetch=50):` for instance. Let me know if it is still unclear.

**Comment by ayuhsya** [comment]
@abinashmeher999 Let me tell you what the problems are:

- If we want to give user the option to specify `num_per_fetch` for every call, then the usage of the `get_linkshere()` function will be different. An instance will have to be created and `next()` would be called on it in a loop **by the user**. Is this acceptable?

- If there are multiple `pages/titles` specified then it is not confirmed that you can get 300 links for all of them in the first call. This is because if you read the response in the case of multiple `pages/titles` ([example](https://en.wikipedia.org/w/api.php?action=query&titles=Main%20Page%7Ctesla&prop=linkshere)) only the links on one of all the specified `pages/titles` is returned and others are put on hold till all the links from the first `page/title` is finished. So it's not possible to get let's say the first 300 links on all the specified `pages/titles` in this case in the first API call itself.

**Comment by abinashmeher999** [comment]
@ayuhsya It would be unwise to have such an API where the user mentions the limit in every fetch. I am having a hard time thinking of such a use case where the user insists on that. So better to let the user mention only one limit which will be used for all further fetches from the API.

For the second case, How do you suggest can the user iterate on all the entries? Only then we can say if that is possible or not. When I thought of lazily fetching entries I had thought of only one page. Because it would be way too confusing when you start iterating over all the entries in a group of pages. Maybe we need better design before we can try the same issue for group of pages.
